cobalt-parser/src/parser/expr/literal.rs

```diff
@@ -0,0 +1,506 @@
+
+            fields.insert(field_name.clone(), expr);
+            field_spans.insert(field_name, field_name_span);
+
+            local_state = atleast_one_field;
+        }
+
+        // ---
+
+        (Box::new(StructLiteralAST::new(span, fields)), errors)
+    }
+
+    /// Going into this function, expect current token to be 'LiteralToken::Str'.
+    pub(crate) fn parse_string_literal(
+        &mut self,
+    ) -> (Option<StringLiteralAST<'src>>, Vec<CobaltError<'src>>) {
+        assert!(self.current_token.is_some());
+
+        let span = self.current_token.unwrap().span;
+        let mut errors = vec![];
+
+        let mut chars: Chars;
+        if let TokenKind::Literal(LiteralToken::Str(s)) = self.current_token.unwrap().kind {
+            chars = s.chars();
+        } else {
+            self.next();
+            errors.push(CobaltError::ExpectedFound {
+                ex: "string literal",
+                found: ParserFound::Str(self.current_token.unwrap().kind.to_string()),
+                loc: span,
+            });
+            return (None, errors);
+        }
+
+        self.next();
+
+        // ---
+
+        let mut cbi_s: Vec<CharBytesIterator> = vec![];
+        loop {
+            let cbi = match chars.next() {
+                None => break,
+                Some('\\') => {
+                    match chars.next() {
+                        None => break,
+                        Some('0') => CharBytesIterator::from_u8(0x00),
+                        Some('n') => CharBytesIterator::from_u8(0x0a),
+                        Some('r') => CharBytesIterator::from_u8(0x0d),
+                        Some('t') => CharBytesIterator::from_u8(0x09),
+                        Some('v') => CharBytesIterator::from_u8(0x0a),
+                        Some('b') => CharBytesIterator::from_u8(0x08),
+                        Some('e') => CharBytesIterator::from_u8(0x1b),
+                        Some('a') => CharBytesIterator::from_u8(0x07),
+                        Some('c') => {
+                            // --- Next are exactly two hex digits.
+
+                            let mut digit1 = chars.next();
+                            let mut digit2 = chars.next();
+                            if digit1.is_none() || digit2.is_none() {
+                                errors.push(CobaltError::ExpectedFound {
+                                    ex: "two hex digits",
+                                    found: ParserFound::Str("something else".to_string()),
+                                    loc: SourceSpan::from((self.cursor.index, 2)),
+                                });
+                                digit1 = Some('0');
+                                digit2 = Some('0');
+                            }
+
+                            let digit1 = digit1.unwrap();
+                            let digit2 = digit2.unwrap();
+
+                            let mut buf = [0; 8]; // UTF-8 encoded chars can be up to 4 bytes each
```

Comment:
Creating a new string, even on stack, is unnecessary. char::to_digit(16) handles this much more cleanly and is faster.

Reply:
I'm a little confused on why this would work... don't we want to parse the entire hex literal to a number (`u8` if possible, or `u32`)? How would we achieve that easier by getting the value of the individual digits?